### PR TITLE
Base Structure

### PR DESCRIPTION
**Pull request by mheon**
Structure for us to build on

### CODE REVIEW

**Comment by mheon** [comment]
My indentation is completely screwed up, let me fix...

**Comment by mheon** [comment]
Rustfmt appears to have resolved the issue

File: src/firewall/firewalld.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    conn: Connection,
+}
+
+pub fn new() -> Result<Box<dyn firewall::FirewallDriver>, Box<dyn Error>> {
+    bail!("not yet implemented")
+}
+
+impl firewall::FirewallDriver for FirewallD {
+    fn setup_network(&self, _net: types::Network) -> Result<(), Box<dyn Error>> {
+        bail!("not yet implemented");
```
Comment (cgwalters) [inline src/firewall/firewalld.rs:18]: I use `todo!()` for this

File: src/lib.rs [hunk 2]
```diff
@@ -1,9 +1,13 @@
+extern crate env_logger;
```
Comment (cgwalters) [inline src/lib.rs:1]: These generally aren't necessary anymore in Rust 2018 edition.

File: src/firewall/mod.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+pub trait FirewallDriver {
+    // Set up firewall rules for the given network,
+    fn setup_network(&self, net: types::Network) -> Result<(), Box<dyn Error>>;
+    // Tear down firewall rules for the given network.
+    fn teardown_network(&self, net: types::Network) -> Result<(), Box<dyn Error>>;
+
+    // Set up port-forwarding firewall rules for a given container.
+    fn setup_port_forward(
+        &self,
+        container_id: String,
```
Comment (cgwalters) [inline src/firewall/mod.rs:18]: Usually you want function parameters to accept `&str`.

File: src/commands/setup.rs [hunk 4]
```diff
@@ -32,6 +33,28 @@ impl Setup {
             Err(e) => panic!("{}", e),
         };
 
+        let firewall_driver = match firewall::get_supported_firewall_driver() {
+            Ok(driver) => driver,
+            Err(e) => panic!("{}", e.to_string()),
```
Comment (cgwalters) [inline src/commands/setup.rs:38]: Any reason to panic instead of propagating the error?

**Pull request merged**